File: python/lightning-enable-mcp/src/lightning_enable_mcp/price_service.py
"""
Price Service

Fetches BTC/USD price from three independent public sources (CoinGecko,
Coinbase, Kraken) in parallel and returns the first successful response.

Design:
- 60-second cache: keeps price close to spot while absorbing burst traffic
  and avoiding rate-limit pressure on any single source.
- Parallel fetch on cache miss: a slow or failing source cannot hold up the
  others — the first success wins.
- No hardcoded fallback price. If every source fails, the service RAISES
  PriceUnavailableError. A wrong fake price would silently mis-evaluate
  budgets. Better to fail loud.
- Every fetch attempt is logged (source, latency, success/failure).
"""

from __future__ import annotations

import asyncio
import logging
import time
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from decimal import Decimal
from typing import Awaitable, Callable

try:
    import httpx
except ImportError:
    httpx = None  # type: ignore

logger = logging.getLogger("lightning-enable-mcp.price")

# Satoshis per Bitcoin
SATS_PER_BTC = Decimal("100000000")

# Cache duration — short, so price stays close to spot.
CACHE_DURATION = timedelta(seconds=60)

# Per-source timeout. A slow source must not block the others.
PER_SOURCE_TIMEOUT_SECONDS = 5.0


class PriceUnavailableError(Exception):
    """
    Raised when CoinGecko, Coinbase, and Kraken all fail and no recent
    cached value is available. The MCP refuses to fall back to a fake price
    because that would mis-evaluate budgets.
    """


@dataclass(frozen=True)
class PriceSnapshot:
    """A point-in-time price with provenance."""

    btc_usd: Decimal
    source: str
    fetched_at: datetime


class PriceService:
    """
    BTC/USD price service. Fetches from CoinGecko, Coinbase, Kraken in
    parallel; first success wins.

    Sources are intentionally limited to public, unauthenticated APIs that
    are widely trusted. Strike is NOT used here even when configured — it's
    surfaced separately via the explicit get_btc_price tool for Strike-only
    users.
    """

    def __init__(self) -> None:
        if httpx is None:
            raise ImportError(
                "httpx is required for PriceService. Install with: pip install httpx"
            )

        self._lock = asyncio.Lock()
        self._client: httpx.AsyncClient | None = None
        self._snapshot: PriceSnapshot | None = None

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create the shared HTTP client."""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                timeout=PER_SOURCE_TIMEOUT_SECONDS,
                headers={
                    "Accept": "application/json",
                    "User-Agent": "lightning-enable-mcp/price",
                },
            )
        return self._client

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client and not self._client.is_closed:
            await self._client.aclose()
            self._client = None

    async def get_btc_price(self) -> Decimal:
        """
        Get the current BTC/USD price.

        Returns the cached price if it's less than CACHE_DURATION old,
        otherwise fires CoinGecko/Coinbase/Kraken in parallel and returns
        the first successful response. Raises PriceUnavailableError if all
        sources fail.
        """
        snapshot = self._snapshot
        if snapshot is not None and datetime.now(timezone.utc) - snapshot.fetched_at < CACHE_DURATION:
            return snapshot.btc_usd

        async with self._lock:
            # Re-check after acquiring lock — another caller may have refreshed.
            snapshot = self._snapshot
            if snapshot is not None and datetime.now(timezone.utc) - snapshot.fetched_at < CACHE_DURATION:
                return snapshot.btc_usd

            fresh = await self._fetch_first_successful()
            self._snapshot = fresh
            return fresh.btc_usd

    async def sats_to_usd(self, sats: int) -> Decimal:
        """
        Convert satoshis to USD using the current BTC price.
        Raises PriceUnavailableError if the price cannot be fetched.
        """
        btc_price = await self.get_btc_price()
        usd = Decimal(sats) / SATS_PER_BTC * btc_price
        return round(usd, 2)

    async def usd_to_sats(self, usd: Decimal | float | int) -> int:
        """
        Convert USD to satoshis using the current BTC price (rounded up).
        Raises PriceUnavailableError if the price cannot be fetched.
        """
        import math

        btc_price = await self.get_btc_price()
        usd_decimal = Decimal(str(usd))
        sats = usd_decimal / btc_price * SATS_PER_BTC
        return math.ceil(sats)

    def get_last_snapshot(self) -> PriceSnapshot | None:
        """
        Returns the most recent successfully fetched snapshot (price + source +
        timestamp) without triggering a fresh fetch. None if no fetch has
        succeeded yet.
        """
        return self._snapshot

    def get_cached_btc_price(self) -> Decimal:
        """
        Returns the last successfully fetched price for synchronous callers.
        Returns Decimal("0") if no fetch has succeeded yet — callers must
        handle this case rather than relying on a hardcoded fake price.
        """
        if self._snapshot is None:
            return Decimal("0")
        return self._snapshot.btc_usd

    def get_cache_source(self) -> str | None:
        """The source of the cached price, or None if no fetch has succeeded."""
        return self._snapshot.source if self._snapshot else None

    def is_cache_valid(self) -> bool:
        """Whether the cached value is within CACHE_DURATION."""
        if self._snapshot is None:
            return False
        return datetime.now(timezone.utc) - self._snapshot.fetched_at < CACHE_DURATION

    # ── internals ────────────────────────────────────────────────────────────

    async def _fetch_first_successful(self) -> PriceSnapshot:
        """
        Fire CoinGecko, Coinbase, and Kraken in parallel; return the first
        successful PriceSnapshot. Raise PriceUnavailableError if all fail.
        """
        attempts: list[tuple[str, Callable[[], Awaitable[Decimal]]]] = [
            ("CoinGecko", self._fetch_coingecko),
            ("Coinbase", self._fetch_coinbase),
            ("Kraken", self._fetch_kraken),
        ]
        tasks = {asyncio.create_task(self._try(source, fetch)): source for source, fetch in attempts}

        winner: PriceSnapshot | None = None
        try:
            for completed in asyncio.as_completed(tasks):
                result = await completed
                if result is not None:
                    winner = result
                    break
        finally:
            for task in tasks:
                if not task.done():
                    task.cancel()
            # Drain cancelled tasks so they don't surface as warnings.
            for task in tasks:
                try:
                    await task
                except (asyncio.CancelledError, Exception):
                    pass

        if winner is not None:
            logger.info(
                "BTC price fetched: $%s from %s",
                f"{winner.btc_usd:,.2f}",
                winner.source,
            )
            return winner

        message = (
            "BTC price unavailable: CoinGecko, Coinbase, and Kraken all failed. "
            "Cannot evaluate budget safely."
        )
        logger.error(message)
        raise PriceUnavailableError(message)

    async def _try(
        self,
        source: str,
        fetch: Callable[[], Awaitable[Decimal]],
    ) -> PriceSnapshot | None:
        """Run a single source fetch with timing and error logging."""
        started = time.monotonic()
        try:
            price = await fetch()
            elapsed_ms = (time.monotonic() - started) * 1000
            if price <= 0:
                logger.warning(
                    "BTC price fetch from %s returned non-positive value %s after %.1fms",
                    source, price, elapsed_ms,
                )
                return None
            logger.debug("BTC price fetched from %s: $%s in %.1fms", source, price, elapsed_ms)
            return PriceSnapshot(
                btc_usd=price,
                source=source,
                fetched_at=datetime.now(timezone.utc),
            )
        except asyncio.CancelledError:
            # Sibling source already won — not a failure.
            raise
        except Exception as ex:
            elapsed_ms = (time.monotonic() - started) * 1000
            logger.warning(
                "BTC price fetch from %s failed after %.1fms: %s",
                source, elapsed_ms, ex,
            )
            return None

    async def _fetch_coingecko(self) -> Decimal:
        client = await self._get_client()
        response = await client.get(
            "https://api.coingecko.com/api/v3/simple/price",
            params={"ids": "bitcoin", "vs_currencies": "usd"},
        )
        response.raise_for_status()
        data = response.json()
        if "bitcoin" in data and "usd" in data["bitcoin"]:
            return Decimal(str(data["bitcoin"]["usd"]))
        raise ValueError("CoinGecko response missing bitcoin.usd")

    async def _fetch_coinbase(self) -> Decimal:
        client = await self._get_client()
        response = await client.get("https://api.coinbase.com/v2/prices/BTC-USD/spot")
        response.raise_for_status()
        data = response.json()
        if "data" in data and "amount" in data["data"]:
            return Decimal(str(data["data"]["amount"]))
        raise ValueError("Coinbase response missing data.amount")

    async def _fetch_kraken(self) -> Decimal:
        client = await self._get_client()
        response = await client.get(
            "https://api.kraken.com/0/public/Ticker",
            params={"pair": "XBTUSD"},
        )
        response.raise_for_status()
        data = response.json()

        errors = data.get("error") or []
        if errors:
            raise ValueError(f"Kraken error: {errors[0]}")

        result = data.get("result") or {}
        # Pair key is typically "XXBTZUSD" but may vary.
        for pair_data in result.values():
            close = pair_data.get("c")
            if isinstance(close, list) and close:
                return Decimal(str(close[0]))

        raise ValueError("Kraken response missing close price")


# Module-level singleton for convenience.
_default_service: PriceService | None = None


def get_price_service() -> PriceService:
    """Get the default price service singleton."""
    global _default_service
    if _default_service is None:
        _default_service = PriceService()
    return _default_service


async def get_btc_price() -> Decimal:
    """Get current BTC/USD price using the default service."""
    return await get_price_service().get_btc_price()


async def sats_to_usd(sats: int) -> Decimal:
    """Convert satoshis to USD using the default service."""
    return await get_price_service().sats_to_usd(sats)


async def usd_to_sats(usd: Decimal | float | int) -> int:
    """Convert USD to satoshis using the default service."""
    return await get_price_service().usd_to_sats(usd)

File: dotnet/src/LightningEnable.Mcp/Services/PriceService.cs
using System.Globalization;
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace LightningEnable.Mcp.Services;

/// <summary>
/// Service for converting between USD and satoshis using current BTC/USD price.
///
/// Design:
/// • Three independent public sources: CoinGecko, Coinbase, Kraken.
/// • On cache miss, all three are queried in parallel and the first successful
///   response wins (resilient to a single slow/failing source).
/// • Cache duration is short (60 seconds) so price stays close to spot while
///   absorbing burst traffic and avoiding rate-limit pressure.
/// • If every source fails, the service THROWS — there is no hardcoded
///   fallback price. A wrong fake price would silently mis-evaluate budgets.
/// • Every fetch attempt is logged (source, latency, success/failure).
/// </summary>
public interface IPriceService
{
    /// <summary>
    /// Gets the current BTC/USD price. Throws if all sources fail and no
    /// recent cached value is available.
    /// </summary>
    Task<decimal> GetBtcPriceAsync(CancellationToken cancellationToken = default);

    /// <summary>
    /// Converts USD to satoshis using the current BTC price.
    /// </summary>
    Task<long> UsdToSatsAsync(decimal usd, CancellationToken cancellationToken = default);

    /// <summary>
    /// Converts satoshis to USD using the current BTC price.
    /// </summary>
    Task<decimal> SatsToUsdAsync(long sats, CancellationToken cancellationToken = default);

    /// <summary>
    /// Returns the most recent successfully fetched price along with the
    /// source name and timestamp. Returns null if no fetch has succeeded yet.
    /// Does not trigger a fresh fetch — use GetBtcPriceAsync for that.
    /// </summary>
    PriceSnapshot? GetLastSnapshot();
}

/// <summary>
/// A point-in-time price snapshot with provenance.
/// </summary>
public sealed record PriceSnapshot(decimal BtcUsd, string Source, DateTime FetchedAtUtc);

/// <summary>
/// Thrown when every price source fails and no recent cached value is available.
/// </summary>
public sealed class PriceUnavailableException : InvalidOperationException
{
    public PriceUnavailableException(string message) : base(message) { }
    public PriceUnavailableException(string message, Exception innerException)
        : base(message, innerException) { }
}

public class PriceService : IPriceService
{
    private static readonly TimeSpan CacheDuration = TimeSpan.FromSeconds(60);
    private static readonly TimeSpan PerSourceTimeout = TimeSpan.FromSeconds(5);

    private readonly HttpClient _httpClient;
    private readonly ILogger<PriceService>? _logger;
    private readonly SemaphoreSlim _refreshLock = new(1, 1);

    private PriceSnapshot? _cached;

    public PriceService(HttpClient httpClient, ILogger<PriceService>? logger = null)
    {
        _httpClient = httpClient;
        _logger = logger;
    }

    public async Task<decimal> GetBtcPriceAsync(CancellationToken cancellationToken = default)
    {
        // Fast path: cache is fresh.
        var snapshot = _cached;
        if (snapshot != null && DateTime.UtcNow - snapshot.FetchedAtUtc < CacheDuration)
        {
            return snapshot.BtcUsd;
        }

        await _refreshLock.WaitAsync(cancellationToken);
        try
        {
            // Re-check after acquiring lock — another caller may have refreshed.
            snapshot = _cached;
            if (snapshot != null && DateTime.UtcNow - snapshot.FetchedAtUtc < CacheDuration)
            {
                return snapshot.BtcUsd;
            }

            var fresh = await FetchFirstSuccessfulAsync(cancellationToken);
            _cached = fresh;
            return fresh.BtcUsd;
        }
        finally
        {
            _refreshLock.Release();
        }
    }

    public async Task<long> UsdToSatsAsync(decimal usd, CancellationToken cancellationToken = default)
    {
        var price = await GetBtcPriceAsync(cancellationToken);
        var sats = usd / price * 100_000_000m;
        return (long)Math.Ceiling(sats);
    }

    public async Task<decimal> SatsToUsdAsync(long sats, CancellationToken cancellationToken = default)
    {
        var price = await GetBtcPriceAsync(cancellationToken);
        return Math.Round(sats / 100_000_000m * price, 2);
    }

    public PriceSnapshot? GetLastSnapshot() => _cached;

    /// <summary>
    /// Fires CoinGecko, Coinbase, and Kraken in parallel and returns the first
    /// successful result. Throws PriceUnavailableException if all fail.
    /// </summary>
    private async Task<PriceSnapshot> FetchFirstSuccessfulAsync(CancellationToken cancellationToken)
    {
        using var attemptCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);

        var tasks = new List<Task<PriceSnapshot?>>
        {
            TryFetchAsync("CoinGecko", FetchCoinGeckoAsync, attemptCts.Token),
            TryFetchAsync("Coinbase",  FetchCoinbaseAsync,  attemptCts.Token),
            TryFetchAsync("Kraken",    FetchKrakenAsync,    attemptCts.Token)
        };

        var failures = new List<string>();
        while (tasks.Count > 0)
        {
            var completed = await Task.WhenAny(tasks);
            tasks.Remove(completed);

            var snapshot = await completed;
            if (snapshot != null)
            {
                // First success wins — cancel the rest to free sockets.
                attemptCts.Cancel();
                _logger?.LogInformation(
                    "BTC price fetched: ${Price} from {Source}",
                    snapshot.BtcUsd,
                    snapshot.Source);
                return snapshot;
            }
        }

        var message =
            "BTC price unavailable: CoinGecko, Coinbase, and Kraken all failed. " +
            "Cannot evaluate budget safely.";
        _logger?.LogError("{Message}", message);
        throw new PriceUnavailableException(message);
    }

    private async Task<PriceSnapshot?> TryFetchAsync(
        string source,
        Func<CancellationToken, Task<decimal>> fetch,
        CancellationToken cancellationToken)
    {
        var startedAt = DateTime.UtcNow;
        using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutCts.CancelAfter(PerSourceTimeout);

        try
        {
            var price = await fetch(timeoutCts.Token);
            if (price <= 0)
            {
                _logger?.LogWarning(
                    "BTC price fetch from {Source} returned non-positive value {Price}",
                    source,
                    price);
                return null;
            }

            return new PriceSnapshot(price, source, DateTime.UtcNow);
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            // Caller cancelled (a sibling source already succeeded) — not a failure.
            return null;
        }
        catch (OperationCanceledException)
        {
            _logger?.LogWarning(
                "BTC price fetch from {Source} timed out after {ElapsedMs}ms",
                source,
                (DateTime.UtcNow - startedAt).TotalMilliseconds);
            return null;
        }
        catch (Exception ex)
        {
            _logger?.LogWarning(
                ex,
                "BTC price fetch from {Source} failed after {ElapsedMs}ms: {Message}",
                source,
                (DateTime.UtcNow - startedAt).TotalMilliseconds,
                ex.Message);
            return null;
        }
    }

    private async Task<decimal> FetchCoinGeckoAsync(CancellationToken cancellationToken)
    {
        using var response = await _httpClient.GetAsync(
            "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=usd",
            cancellationToken);
        response.EnsureSuccessStatusCode();

        var json = await response.Content.ReadAsStringAsync(cancellationToken);
        using var doc = JsonDocument.Parse(json);
        if (doc.RootElement.TryGetProperty("bitcoin", out var btc) &&
            btc.TryGetProperty("usd", out var usd))
        {
            return usd.GetDecimal();
        }
        throw new InvalidOperationException("CoinGecko response missing bitcoin.usd");
    }

    private async Task<decimal> FetchCoinbaseAsync(CancellationToken cancellationToken)
    {
        using var response = await _httpClient.GetAsync(
            "https://api.coinbase.com/v2/prices/BTC-USD/spot",
            cancellationToken);
        response.EnsureSuccessStatusCode();

        var json = await response.Content.ReadAsStringAsync(cancellationToken);
        using var doc = JsonDocument.Parse(json);
        if (doc.RootElement.TryGetProperty("data", out var data) &&
            data.TryGetProperty("amount", out var amount))
        {
            var raw = amount.GetString();
            if (decimal.TryParse(raw, NumberStyles.Number, CultureInfo.InvariantCulture, out var price))
            {
                return price;
            }
            throw new InvalidOperationException($"Coinbase amount unparseable: {raw}");
        }
        throw new InvalidOperationException("Coinbase response missing data.amount");
    }

    private async Task<decimal> FetchKrakenAsync(CancellationToken cancellationToken)
    {
        using var response = await _httpClient.GetAsync(
            "https://api.kraken.com/0/public/Ticker?pair=XBTUSD",
            cancellationToken);
        response.EnsureSuccessStatusCode();

        var json = await response.Content.ReadAsStringAsync(cancellationToken);
        using var doc = JsonDocument.Parse(json);

        // Kraken response shape: { "error":[], "result":{ "XXBTZUSD":{ "c":["76793.00","..."], ... } } }
        if (doc.RootElement.TryGetProperty("error", out var error) &&
            error.ValueKind == JsonValueKind.Array &&
            error.GetArrayLength() > 0)
        {
            var firstError = error[0].GetString() ?? "unknown";
            throw new InvalidOperationException($"Kraken error: {firstError}");
        }

        if (!doc.RootElement.TryGetProperty("result", out var result))
        {
            throw new InvalidOperationException("Kraken response missing result");
        }

        // The pair key may be "XXBTZUSD" or "XBTUSD" depending on Kraken's mood.
        foreach (var pair in result.EnumerateObject())
        {
            if (pair.Value.TryGetProperty("c", out var close) &&
                close.ValueKind == JsonValueKind.Array &&
                close.GetArrayLength() > 0)
            {
                var raw = close[0].GetString();
                if (decimal.TryParse(raw, NumberStyles.Number, CultureInfo.InvariantCulture, out var price))
                {
                    return price;
                }
                throw new InvalidOperationException($"Kraken close price unparseable: {raw}");
            }
        }

        throw new InvalidOperationException("Kraken response missing close price");
    }
}
